### PR TITLE
Da story 30

### PR DESCRIPTION
**Pull request by d-atkins**
## Description
Completes User Story #30

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which maintains existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Notes
Adds order cancellation - cancelling an order changes the status of the order, all of its order items, and restocks any fulfilled items' inventories
- functionality to cancel an order lives in profile::orders controller - feels a little weird but "cancelling" is the only way for a user to update the status of an order as far as I know
  - adds a `cancel` model method to `order` and `restock` method to `item_order`
    - currently uses an `each` block to iterate, not sure how to avoid that
```
User Story 30, User cancels an order

As a registered user
When I visit an order's show page
I see a button or link to cancel the order 
When I click the cancel button for an order, the following happens:
- Each row in the "order items" table is given a status of "unfulfilled"
- The order itself is given a status of "cancelled"
- Any item quantities in the order that were previously fulfilled have their quantities returned to their respective merchant's inventory for that item.
- I am returned to my profile page
- I see a flash message telling me the order is now cancelled
- And I see that this order now has an updated status of "cancelled"
```

## RSpec Results
```
161 examples, 0 failures

Coverage report generated for RSpec - 2480 / 2480 LOC (100.0%) covered.
```
*hard to test in rails s until story 50 is implemented - have to change an item_order's status to "fulfilled" in rails c

### CODE REVIEW

File: app/models/order.rb [hunk 1]
```diff
@@ -14,7 +18,11 @@ def item_count
...
   end
 
-  def self.by_status
-    order(status: :asc)
+  def cancel
+    update(status: "cancelled")
+    item_orders.where(status: "fulfilled").each do |item_order|
+      item_order.item.increment!(:inventory, item_order.quantity)
+    end
+    item_orders.update(status: "unfulfilled")
```
Comment (iEv0lv3) [inline app/models/order.rb]: This is a complicated one to solve from a pure AR perspective. Maybe we could go over it together in Postico using the fancy button in the top-left?

File: spec/features/admin/users/user_profile_page_spec.rb [hunk 2]
```diff
@@ -3,8 +3,8 @@
 RSpec.describe 'As an admin', type: :feature do
   describe "When I visit a user's profile page" do
     before :each do
-      @user = create(:admin_user)
```
Comment (iEv0lv3) [inline spec/features/admin/users/user_profile_page_spec.rb:6]: Thank you for cleaning these up from my PR!

File: app/controllers/profile/orders_controller.rb [hunk 3]
```diff
@@ -29,6 +29,16 @@ def create
     end
   end
 
+  def update
+    order = Order.find(params[:id])
+    unless order.cancelled? || order.shipped?
+      order.cancel
```
Comment (iEv0lv3) [inline app/controllers/profile/orders_controller.rb:35]: I like that you used a model method here, and the naming convention makes it clear what it does.

File: app/models/order.rb [hunk 4]
```diff
@@ -6,6 +6,10 @@ class Order <ApplicationRecord
   has_many :item_orders
   has_many :items, through: :item_orders
 
+  def self.by_status
+    order(status: :asc)
+  end
```
Comment (iEv0lv3) [inline app/models/order.rb:11]: I didn't know how to feel about changing the `order.status` enum default order as a solution (from one of my PRs), but looking at this query makes me feel very okay about it 👍 I like this class method!

**Review by kathleen-crowe — APPROVED** [review]

**Pull request merged**